### PR TITLE
[SM-670] Restrict UI actions based on user permission

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to restrict UI actions that the current user doesn't have permission to preform related to projects.

## Code changes

- apps/web/src/locales/en/messages.json

New messages for no items project → secrets tab and bulk confirmation dialog.

- bitwarden_license/bit-web/src/app/secrets-manager/models/view/project-list.view.ts
bitwarden_license/bit-web/src/app/secrets-manager/models/view/project.view.ts
bitwarden_license/bit-web/src/app/secrets-manager/projects/models/responses/project-list-item.response.ts
bitwarden_license/bit-web/src/app/secrets-manager/projects/models/responses/project.response.ts
bitwarden_license/bit-web/src/app/secrets-manager/projects/project.service.ts
bitwarden_license/bit-web/src/app/secrets-manager/projects/project/project.component.ts

Updating response models to contain current user's permission.

- bitwarden_license/bit-web/src/app/secrets-manager/projects/project/project-secrets.component.html
bitwarden_license/bit-web/src/app/secrets-manager/projects/project/project-secrets.component.ts

Grab current project to check permissions and display a custom no items message if the user only has read access.

- bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts

Now that we are including permissions in server responses; on access removal going from read/write → read-only, we want to invoke an access policy refresh from the accessPolicyService. 

This will force the project list to refresh and give us up-to-date permissions on the projects.

- bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.html
bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts

Added UI restrictions based on access. 
Wrote some custom logic that fires on selection change to determine if the user has write access to the selections or not.
Based on the flag, we hide the bulk menu since it only has the delete action.

- bitwarden_license/bit-web/src/app/secrets-manager/projects/projects/projects.component.ts

Launch the bulk confirmation dialog when the user has a combination of read/write and read-only access.

- bitwarden_license/bit-web/src/app/secrets-manager/shared/dialogs/bulk-confirmation-dialog.component.html
bitwarden_license/bit-web/src/app/secrets-manager/shared/dialogs/bulk-confirmation-dialog.component.ts

New bulk confirmation dialog used to display to the user what items will not be deleted due to permission restrictions. 

- bitwarden_license/bit-web/src/app/secrets-manager/shared/sm-shared.module.ts

Expose the bulk confirmation dialog in the sm-shared module.

- bitwarden_license/bit-web/src/app/secrets-manager/secrets/dialog/secret-dialog.component.ts

Have project selection be permission aware.

### Read only 
image

### Write

image

### Bulk delete projects when the user has write access to all projects

https://user-images.githubusercontent.com/43214426/231276263-24915e47-6c4a-4d0e-a3ae-6961f11b7f10.mov

### Bulk delete projects when the user has a combination of read-only and write access to projects

https://user-images.githubusercontent.com/43214426/231276358-da08e33c-60a3-41a2-8afc-2e788c216406.mov

### Bulk delete projects when the user has only read-only access to projects

https://user-images.githubusercontent.com/43214426/231276467-92f0cd54-fa66-4ea2-80ff-460802d2eb3f.mov

### Project user has write access – no secrets 

image

### Project User has read-only access – no secrets

image

### Project user has write access 

image

### Project user has read-only access

image

### Before 

image

### After 

image

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by coltonhurst — CHANGES_REQUESTED** [review]
Looks pretty good so far, however Export for projects is broken. We may need to check Import also.

**Review by coltonhurst — CHANGES_REQUESTED** [review]
Looks like the above has been fixed in the [equivalent server PR](https://github.com/bitwarden/server/pull/2822). Thank you!

When creating a secret, we probably shouldn't show projects in the project dropdown that the user only has read access to. I know the action results in an error, but it might be good UX to just not show it. Any thoughts?

pic

**Comment by Thomas-Avery** [comment]
> Looks like the above has been fixed in the [equivalent server PR](https://github.com/bitwarden/server/pull/2822). Thank you!
> 
> When creating a secret, we probably shouldn't show projects in the project dropdown that the user only has read access to. I know the action results in an error, but it might be good UX to just not show it. Any thoughts?
> 
> pic

Good idea I've added that.

**Review by coltonhurst — APPROVED** [review]
Looks good, thank you!

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/dialog/secret-dialog.component.ts [hunk 1]
```diff
@@ -91,10 +90,21 @@ export class SecretDialogComponent implements OnInit {
     this.loading = false;
 
     if (secret.write) {
+      this.projects = this.projects.filter((p) => p.write);
```
Comment (willmartian) [inline bitwarden_license/bit-web/src/app/secrets-manager/secrets/dialog/secret-dialog.component.ts:93]: Why are we calling `this.projects = this.projects.filter((p) => p.write)` in both `loadData` and `loadProjects`? I'm a little confused here

**Review by willmartian — CHANGES_REQUESTED** [review]
Good job here. A couple small comments:

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/dialog/secret-dialog.component.ts (hunk 1 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets/dialog/secret-dialog.component.ts]: I cleaned it up https://github.com/bitwarden/clients/pull/5090/commits/dcadd7c367ef8c6c3b44f99ee2ebb1c4e6b81af2

The idea is, if we have a read-only secret, we need to load all the projects so the read only project association is displayed correctly.

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts [hunk 2]
```diff
@@ -43,10 +43,12 @@ export class ProjectsListComponent implements OnDestroy {
...
     private i18nService: I18nService,
     private platformUtilsService: PlatformUtilsService
-  ) {
-    this.selection.changed
-      .pipe(takeUntil(this.destroy$))
-      .subscribe((_) => this.onProjectCheckedEvent.emit(this.selection.selected));
+  ) {}
+
+  ngOnInit(): void {
+    this.selection.changed.pipe(takeUntil(this.destroy$)).subscribe((_) => {
```
[the remark this answers is not in the document]
Comment (willmartian) [reply bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts]: We don't need the `ngOnInit`. 

On line 42, after we initialize `selection`, we can initialize `hasWriteAccessOnSelected$`:

```ts
selection = new SelectionModel<string>(true, []);
protected hasWriteAccessOnSelected$ = this.selection.changed.pipe(map(_ => this.selectedHasWriteAcess()))
```

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets/dialog/secret-dialog.component.ts (hunk 1 above)
Comment (willmartian) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets/dialog/secret-dialog.component.ts]: Gotcha, thank you 👍

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts]: Take two https://github.com/bitwarden/clients/pull/5090/commits/96892db91fe6b239fed59e8b8478a1f32e7a62fc

[the remark this answers is not in the document]
Comment (willmartian) [reply bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts]: Perfect!

**Review by willmartian — APPROVED** [review]
Nice job! 🌈

**Pull request merged**